sympy/printing/numpy.py

```diff
@@ -290,6 +290,7 @@ def _print_NDimArray(self, expr):
     'gamma': 'gamma',
     'loggamma': 'gammaln',
     'digamma': 'psi',
+    'polygamma': 'polygamma',
```

Comment:
Why is this necessary now if it wasn't necessary before? Has something changed in scipy?

Reply:
I don't actually know. `lambdify` uses `exec` with a namespace which I think contains `__all__` from the imported module. So I guess either the logic in `lambdify` changed or the exported functions in SciPy changed?

Reply:
On master it looks like this:
```
Created with lambdify. Signature:                                                                                                                                      
                                                                                                                                                                       
func(x, y)                                                                                                                                                             
                                                                                                                                                                       
Expression:                                                                                                                                                            
                                                                                                                                                                       
polygamma(x, y)                                                                    
                                                                                                                                                                       
Source code:                      
                                                                                   
def _lambdifygenerated(x, y):                                                      
    return polygamma(x, y)                                                         
                                                                                   
                                                                                   
Imported modules:       

```

and I expected the function call to be qualified when using this branch, but actually it is not... So now I'm just confused.

Reply:
I've been struggling to build scipy myself to test this out. Does `from sympy import *` import polygamma with scipy main? Is that what changed since the 1.9.1 release?

Reply:
Part of the idea of testing with upstream git branches is not just so that SymPy can detect changes and respond accordingly but also to detect regressions in the upstream codebases that might impact SymPy. In other words before changing anything in SymPy we should establish whether this looks like an intentional change in SciPy and if it should be considered a bug in SciPy then it should be reported there so it can be fixed before their next release.

Reply:
Makes sense. I can't seem to reproduce this using Python 3.11 and SciPy-git built a few weeks back:
```
Python 3.11.0b3 (main, Jun 16 2022, 13:59:02) [GCC 12.0.1 20220319 (experimental) [master r12-7719-g8ca61ad148f]]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import scipy

In [2]: from sympy import *

In [3]: from sympy.abc import x,y

In [4]: f = lambdify([x,y], polygamma(x,y))

In [5]: print(f.__doc__)
Created with lambdify. Signature:

func(x, y)

Expression:

polygamma(x, y)

Source code:

def _lambdifygenerated(x, y):
    return polygamma(x, y)


Imported modules:



In [6]: f(1,0.5)
Out[6]: array(4.9348022)

In [7]: scipy.__version__
Out[7]: '1.10.0.dev0+1303.1ef52c6'
```

Reply:
> SciPy-git built a few weeks back:

Looking back through CI the test started failing 3 days ago so we need a more recent SciPy.

Reply:
Thanks, I'll recompile current SciPy master and see what that gets us.

Reply:
Thanks. It could also be a change in numpy. The setup can be seen here:
https://github.com/sympy/sympy/blob/b1f35f8daf5efa9f801f905cf1037000d69b0278/.github/workflows/runtests.yml#L186-L215

Reply:
Alright, so using SciPy master I can reproduce the failure:
```
Python 3.11.0b3 (main, Jun 16 2022, 13:59:02) [GCC 12.0.1 20220319 (experimental) [master r12-7719-g8ca61ad148f]]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import sympy

In [2]: import scipy

In [3]: from sympy.abc import x,y

In [5]: f = sympy.lambdify([x,y], sympy.polygamma(x,y))

In [6]: print(f.__doc__)
Created with lambdify. Signature:

func(x, y)

Expression:

polygamma(x, y)

Source code:

def _lambdifygenerated(x, y):
    return polygamma(x, y)


Imported modules:



In [7]: f(1,0.5)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Input In [7], in <cell line: 1>()
----> 1 f(1,0.5)

File <lambdifygenerated-1>:2, in _lambdifygenerated(x, y)
      1 def _lambdifygenerated(x, y):
----> 2     return polygamma(x, y)

NameError: name 'polygamma' is not defined

In [8]: scipy
Out[8]: <module 'scipy' from '/opt/cpython-3.11.0b3-release/lib/python3.11/site-packages/scipy/__init__.py'>

In [9]: scipy.__version__, sympy.__version__
Out[9]: ('1.10.0.dev0+1668.ac729b8', '1.10.1')
```

Reply:
And using my branch in this PR does not solve this issue.

Reply:
Can you bisect to find the relevant scipy commit?

Reply:
Manually bisecting SciPy commits:
b5e854576b39c784b2c868bd6835b378845511bf OK
d58e74655e1129d5d18f034b5a0219dba90368d3 bad

Looking in scipy/scipy#15607 I see L68-L82 in [`scipy/datasets/__init__.py`](https://github.com/scipy/scipy/pull/15607/files#diff-89f9e47b57fc95d026c7ba70e5915600d02abb9b62a26b2b381d817ac0581160) which looks suspicious.

And indeed: running `pip install pooch` fixes our issue.